ada_url/ada_c.h

```diff
@@ -51,6 +51,7 @@ bool ada_can_parse_with_base(const char* input, size_t input_length,
 
 void ada_free(ada_url result);
 void ada_free_owned_string(ada_owned_string owned);
+ada_url ada_copy(ada_url input);
```

Comment:
Can you override copy operator in python?

Can you add a test similar to what I did on the c tests mainly copying an instance and mutating copied version making sure the initial one isn't mutated?

Reply:
Added in 7e2445ba18939baf9df098477ba9b7f0a3b6e483, but due to how deep copies work in Python, this probably doesn't provide a speed boost relative to constructing a new object.

Reply:
Yes, but the main issue is not around copying it, but with pointers. Since the underlying `ada_url` pointer can not be deep copied, we need `ada_copy` function to trigger the url_aggregator copy constructor.

Reply:
I think we're in agreement! The `__deepcopy__` method I added uses `ada_copy` to ensure that we get a new instance of `ada_url`.

However, I think in Python this will not be faster than `copied_url = URL(original_url.href)`.

Reply:
Nice! I just wanted to clarify the reason for ada_copy. 

Actually; This will trigger the parser and will actually be slower. Because we already know the components of the URL and there is no need to re-run the parser again. I'm 99% sure ada_copy will be faster, but will be a lot happier if I'm mistaken!

Reply:
You were right!

```python
from copy import deepcopy
from timeit import timeit

from ada_url import URL

stmt_old = """
url_obj = URL('https://example.org')
new_ur = URL(url_obj.href)
"""

stmt_new = """
url_obj = URL('https://example.org')
new_url_obj = deepcopy(url_obj)
"""

globals_ = {'URL': URL, 'deepcopy': deepcopy}
old_time = timeit(stmt_old, number=100000, globals=globals_)
new_time = timeit(stmt_new, number=100000, globals=globals_)
print(old_time, new_time, sep='\t')
# 0.45869977000984363	0.3362298179999925
```

I incorrectly assumed that the Python copying overhead would be worse than going back to the C parser.

Reply:
Can you add a test where you create a url and later do a shallow copy and delete the initial url, and checking if the shallowed copied url performs correctly?

Reply:
The Python GC won't collect the old URL object while there's still a reference to it, alas.